staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,268 @@
+	val, _ := m.storageVersionUpdateInfoMap.LoadOrStore(crd.Name, &storageVersionUpdateInfo{})
+	svInfo := val.(*storageVersionUpdateInfo)
+	svInfo.crd = crd
+	if processedCh != nil {
+		svInfo.processedCh = processedCh
+	}
+	m.enqueue(crd.Name)
+}
+
+// enqueue adds the CRD name to the SV upadte queue.
+func (m *Manager) enqueue(crdName string) {
+	if crdName == "" {
+		return
+	}
+	klog.V(4).Infof("enqueueing crd %s for storageversion update", crdName)
+	m.queue.Add(crdName)
+}
+
+// Sync runs a goroutine over the SV updater queue to indefinitely
+// process any queued storageversion updates unless stopCh is invoked.
+func (m *Manager) Sync(stopCh <-chan struct{}, workers int) {
+	defer m.queue.ShutDownWithDrain()
+	klog.V(2).Infof("starting storageversion sync loop")
+	for i := 0; i < workers; i++ {
+		go wait.Until(func() {
+			m.worker()
+		}, workerFrequency, stopCh)
+	}
+
+	<-stopCh
+}
+
+// DeleteTeardownCountInfo deletes specified key from the sync.Map.
+func (m *Manager) DeleteSVUpdateInfo(crdName string) {
+	m.storageVersionUpdateInfoMap.Delete(crdName)
+}
+
+func (m *Manager) worker() {
+	for m.processLatestUpdateFor() {
+	}
+}
+
+func (m *Manager) processLatestUpdateFor() bool {
+	ctx := context.TODO()
+	key, quit := m.queue.Get()
+	defer m.queue.Done(key)
+	if quit {
+		return false
+	}
+
+	// Fetch the latest CRD from storageVersionUpdateInfoMap to update the storageversion for it.
+	val, ok := m.storageVersionUpdateInfoMap.Load(key)
+	if !ok {
+		klog.V(4).Infof("No pending storageversion update found for crdUID: %s, returning", key)
+		return true
+	}
+	latestSVUpdateInfo := val.(*storageVersionUpdateInfo)
+
+	skip, reason := m.shouldSkipSVUpdate(latestSVUpdateInfo)
+	if skip {
+		klog.V(4).Infof("skipping storagversion update for crd: %s, reason: %s ", latestSVUpdateInfo.crd.Name, reason)
+		return true
+	}
+
+	klog.V(4).Infof("starting storageversion update for crd: %s", latestSVUpdateInfo.crd.Name)
+	err := m.updateStorageVersion(ctx, latestSVUpdateInfo.crd)
+	if err == nil {
+		klog.V(4).Infof("successfully updated storage version for %s", latestSVUpdateInfo.crd.Name)
+		close(latestSVUpdateInfo.processedCh)
+		return true
+	}
+
+	// TODO: indefinitely requeue on error?
+	m.enqueue(latestSVUpdateInfo.crd.Name)
+	return true
+}
+
+func (m *Manager) shouldSkipSVUpdate(svUpdateInfo *storageVersionUpdateInfo) (bool, string) {
+	skip, teardownCount := m.teardownInProgress(svUpdateInfo.crd.Name)
+	if skip {
+		return true, fmt.Sprintf("%d active teardowns", teardownCount)
+	}
+
+	if m.alreadyPublishedLatestSV(svUpdateInfo) {
+		return true, "already at latest storageversion"
+	}
+
+	return false, ""
+}
+
+func (m *Manager) alreadyPublishedLatestSV(svUpdateInfo *storageVersionUpdateInfo) bool {
+	select {
+	case <-svUpdateInfo.processedCh:
+		klog.V(4).Infof("Storageversion is already updated to the latest value for crd: %s, returning", svUpdateInfo.crd.Name)
```

Comment:
Is this situation possible?

Reply:
In the very rare situation that the worker has kicked in and is about to start SV update for a CRD (say the latest version at this time was -v2), and before it begins processing, there is a new version of CRD available (v3) - which gets enqueued in the worker queue. The current worker will now find the latest version from the storageVersionUpdateInfoMap to be v3 and will update the SV to be v3. The queued SV update (v3) will be picked up by the worker in the next iteration.

This may be unnecessary, so ok with removing this condition if you think thats for the better.